py/desitarget/io.py

```diff
@@ -330,6 +330,8 @@ def _bright_or_dark(filename, hdr, data, obscon, mockdata=None):
     if not os.path.exists(newdir):
         os.mkdir(newdir)
     filename = os.path.join(newdir, os.path.basename(filename))
+    # ADM modify the filename with an obscon prefix.
+    filename = filename.replace("targets-", "targets-{}-".format(obscon.lower()))
 
```

Comment:
Are we punting on a `desitarget.io.findfile` routine in this PR? Just askin'.

Reply:
No punting here:

https://github.com/desihub/desitarget/pull/558/commits/1f135c6805acb811b8ff9e5dea71860eaa1ac58d

Reply:
Nice! However, neither `select_targets` nor `io.write_targets` seem to be using this function, and in fact `io.write_targets` is creating the filename on-the-fly rather than using the output of `io.find_target_files`.  It's exactly this behavior that originally broke the integration tests (specifically, `desisim.simexp` couldn't find the output files written by `desitarget.mock.build.targets_truth`.

To clarify this request, take a look at https://github.com/desihub/desitarget/blob/master/py/desitarget/mock/io.py#L30-L51 and how `desitarget.mock.io.findfile` is called by `desitarget.mock.build.targets_truth`: https://github.com/desihub/desitarget/blob/master/py/desitarget/mock/build.py#L971-L974

What both @sbailey and I were thinking is that `io.write_targets` would accept a filename that would not be modified internally, that way the same function could be used both for the data and the mocks.  And indeed, the two little functions in `mock.io` could in fact be moved to `desitarget.io` which could handle the mocks as well as the data.

I hope this is all clear and if not I'm happy to chat off-list.  Also: this request is low-priority relative to some of the other changes in this PR, so we shouldn't necessarily hold up this PR while we hammer this out.

Reply:
Ah, I see...

The existing functions in `desitarget.mock` won't work as-is, as they don't have a `survey` kwarg to distinguish between main survey ("main"), commissioning ("cmx") and SV ("sv1") files and file structures. So, some modification would be required. The mocks also don't write allsky ("supp") targets.

Leave that with me, though and I'll see what I can hack together.

Incidentally, there's a lot of functionality on the data side that isn't in the mocks (randoms, commissioning, supplemental file types, gfas, etc.). So, any resolution just to fix the mock/data integration tests is sort of cheating...just modifying desitarget.io.write_targets() is an oversimplification of the data model and directory structure for all of the target types.

Reply:
@moustakas: I've updated all of the principal `write` and `select` functions on the non-mock side to use `io.find_target_files`. The interface into `io.write_targets` is (necessarily) more complicated than for your mock functions, as I have to deal with `resolve` directories, `cmx/sv/main` directories, `supp` directories, etc.

I think it should be straightforward for the mock functions to call the refactored `select/write` functions, though. Let me know if this looks good enough, and we can merge so you can call the new functions for the mocks. Or, if you'd prefer, hijack this branch and update the mock side of the code directly before merging.